File: aoc_3.jl
function read_bool(path)
    x = split.(readlines(path), "")
    parse.(Bool, reduce(hcat, x))
end

bool2dec(x) = parse(Int, join(convert.(Int, x)), base = 2)

more_ones(x, n) = sum(x) >= (n / 2)

find_rate(x) = [more_ones(row, size(x, 2)) for row in eachrow(x)]

function find_hidden(x)
    i = 1
    while size(x, 2) != 1  # scoping!! for .. eachrow(x)
        row = x[i, :]
        flip = !more_ones(row, length(row))
        x = x[:, xor.(flip, row)]
        i += 1
    end
    x
end

function solve(x; part2 = false)
    f = part2 ? find_hidden : find_rate
    bool2dec(f(x)) * bool2dec(f(.!x))
end

data = read_bool("data/aoc_3")
# data = read_bool("data/aoc_3_test")
solve(data)
solve(data, part2 = true)

File: aoc_3.R
parse_string <- function(x) {
  x <- chartr("01", "FT", x) |> strsplit("")
  x <- do.call(rbind, x)
  mode(x) <- "logical"
  x
}

vec2dec <- function(x)
  as.integer(x) |>
  paste(collapse = "") |>
  strtoi(base = 2)

solve_1 <- function(x) {
  counts <- colSums(x)
  half <- nrow(x) / 2
  vec2dec(counts >= half) * vec2dec(counts < half)
}

find_hidden <- function(x, scrub = FALSE) {
  for (i in seq_len(ncol(x))) {
    ids <- x[, i]
    is_greater <- sum(ids) >= (nrow(x) / 2)
    flip <- xor(scrub, is_greater) # additional flip for <=
    x <- x[xor(flip, ids), ]       # flip if is greater
    if (!is.matrix(x)) break       # have vector when one is left
  }
  x
}

solve_2 <- function(x) {
  ogr <- find_hidden(x) |> vec2dec()
  co2 <- find_hidden(x, scrub = TRUE) |> vec2dec()
  ogr * co2
}

# x <- parse_string(c("00100", "11110", "10110", "10111", "10101", "01111", "00111", "11100", "10000", "11001", "00010", "01010"))
x <- parse_string(readLines("data/aoc_3"))
solve_1(x)
solve_2(x)
